src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java

```diff
@@ -94,6 +96,31 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
         return resultRepository.save(originalResult);
     }
 
+    /**
+     * checks if the user can override an already submitted result. This is only possible if the same tutor overrides before the assessment due date
+     * or if an instructor overrides it.
+     *
+     * If the result does not yet exist or is not yet submitted, this method returns true
+     *
+     * @param existingResult the existing result in case the result is updated (submitted or overridden)
+     * @param exercise the exercise to which the submission and result belong and which potentially includes an assessment due date
+     * @param user the user who initiates a request
+     * @param isAtLeastInstructor whether the given user is an instructor for the given exercise
+     * @return true of the the given user can override a potentially existing result
+     */
+    public boolean isAllowedToOverrideExistingResult(@NotNull Result existingResult, Exercise exercise, User user, boolean isAtLeastInstructor) {
+        // if the assessor is null, the user can save / submit / override the existing result
+        final var isAssessor = existingResult.getAssessor() == null || user.equals(existingResult.getAssessor());
+        if (existingResult.getCompletionDate() == null) {
+            // if the result exists, but was not yet submitted (i.e. completionDate not set), the tutor and the instructor can override, independent of the assessment due date
+            return isAssessor || isAtLeastInstructor;
+        }
+        // if the result was already submitted, the tutor can only override before a potentially existing assessment due date
+        var assessmentDueDate = exercise.getAssessmentDueDate();
+        final var isBeforeAssessmentDueDate = assessmentDueDate != null && ZonedDateTime.now().isBefore(assessmentDueDate);
```

Comment:
What if there is no assessment due date? I noticed this when testing, for exercises without assessment due date, a tutor cannot override his own assessment. The override button can be clicked but the tutor just receives a `403` from the backend then.

Reply:
That's intended. The general idea is that as soon as a student can see the result, it should not change anymore. So if the assessment due date has passed or there is no assessment due date, tutors cannot change the assessment any more after submitting it.

If the buttons are still shown on the client side, we should hide them

Reply:
Ok, makes sense, thanks for clarifying.